server/source/controllers/todos.ts

```diff
@@ -0,0 +1,110 @@
+import { Request, Response } from 'express';
+import { v4 as uuidv4 } from 'uuid';
+import { delay } from '../helpers';
+import { mockTodos } from '../mocks/data';
+import { Todo } from '../types';
+
+
+const getAllTodos = async (req: Request, res: Response) => {
+  await delay(800);
+  let todos: Todo[] = mockTodos;
+
+  return res
+    .status(200)
+    .json({ message: todos, status: 'OK' })
+}
+
+
+const getTodo = async (req: Request, res: Response) => {
+  await delay(800);
+  const id: string = req.params.id;
+  let requestedTodo: Todo | undefined = mockTodos.find(todo => todo.id === id);
+
+  // if the tоdo was not found
+  if (!requestedTodo) {
+    const errorMessage: string = `A todo with id '${id}' was not found`
+
+    return res
+      .status(404)
+      .json({ message: errorMessage, status: 'Error' })
+  }
+
+  // if the tоdo was found
+  return res
+    .status(200)
+    .json({ message: requestedTodo, status: 'OK' })
+}
+
+
+const updateTodo = async (req: Request, res: Response) => {
+  await delay(800);
+  const id: string = req.params.id;
+  const newTitle: string = req.body.title ?? null;
+  const todoToUpdate: Todo | undefined = mockTodos.find(todo => todo.id === id)
+
+  // if the tоdo was not found
+  if (!todoToUpdate) {
+    const errorMessage: string = `A todo with id '${id}' was not found`
+
+    return res
+      .status(404)
+      .json({ message: errorMessage, status: 'Error' })
+  }
+
+  // if the tоdo was found
+  if (todoToUpdate) {
+    todoToUpdate.title = newTitle
+  }
+
+  return res
+    .status(200)
+    .json({ message: todoToUpdate, status: 'OK' })
+}
+
+
+const deleteTodo = async (req: Request, res: Response) => {
+  await delay(800);
+  const id: string = req.params.id;
+  const todoToDelete: Todo | undefined = mockTodos.find(todo => todo.id === id);
```

Comment:
if you type `mockTodos` as array of Todo, then TS will do type inferring(not sure if it's correct term) for us as `undefined | ToDo`

Reply:
![image](https://user-images.githubusercontent.com/23404373/188499624-0232e251-656c-4514-bacc-7432e3478946.png)

As I can see if we remove "or undefined" type, TS will be unsatisfied

Reply:
@skyboyer what do you think about it?